source/codegen/templates/service.cpp.mako

```diff
@@ -203,3 +154,34 @@ ${mako_helper.define_simple_method_body(function_name=function_name, function_da
   }
 } // namespace ${config["namespace_component"]}_grpc
 
+% if any(input_custom_types) or any(output_custom_types):
+namespace nidevice_grpc {
+namespace converters {
+%   for custom_type in custom_types:
+%     if custom_type["name"] in output_custom_types:
+template <>
+void convert_to_grpc(const ${custom_type["name"]}& input, ${namespace_prefix}${custom_type["grpc_name"]}* output) 
+{
+%       for field in custom_type["fields"]: 
+  output->set_${field["grpc_name"]}(input.${field["name"]});
+%       endfor
+}
+
+%     endif
+%     if custom_type["name"] in input_custom_types:
+template <>
+${custom_type["name"]} convert_from_grpc(const ${namespace_prefix}${custom_type["grpc_name"]}& input) 
+{
+  ${custom_type["name"]}* output = new ${custom_type["name"]}();  
```

Comment:
Is there a reason this is `new`d instead of allocating it on the stack and relying on a copy constructor? I'm worried that this will lead to memory leaks when we call this and don't put it in a `RepeatedPtrField`.

(my C++ is pretty rusty though :-( )

Reply:
Ooh, yikes. Yeah, that's a bug. The array implementation of `convert_to_grpc` uses new + AddAllocated to transfer ownership. That's weird but it works. This ... is just a leak. I think in either the array or the scalar case. It's returning a copy of the heap allocated struct and never freeing it.